### PR TITLE
Disallow LIKE patterns with trailing escape

### PR DESCRIPTION
**Pull request by rschu1ze**
Trailing escape (`ab\`) is disallowed in SQL, in standardese:

  "If an escape character is specified, then [...] If there is not a
  partitioning of the string PVC into substrings such that each substring
  has length 1 (one) or 2, no substring of length 1 (one) is the escape
  character ECV, and each substring of length 2 is the escape character
  ECV followed by either the escape character ECV, an <underscore>
  character, or the <percent> character, then an exception condition is
  raised: data exception - invalid escape sequence."

I first thought this is checked already higher up in the stack, at least
for const needles, as single trailing backslashes ('ab\') are rejected,
but then I realized that ClickHouse quotes by default. I.e., double
trailing backslashes ('ab\\\\') are not rejected but when interpreted as
LIKE needle ('ab\\') they should.

### Changelog category (leave one):
- Improvement
- Backward Incompatible Change

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
... LIKE patterns with trailing escape symbol ('\\') are now disallowed (as mandated by the SQL standard)

### CODE REVIEW

**Review by alexey-milovidov — APPROVED** [review]
LGTM

**Comment by alexey-milovidov** [comment]
Although it's slightly controversial, because:
1. We are using C-style escaped strings (unlike SQL standard, like MySQL, unlike Postgres, but like Postgres's `e'...'` literals), so this already differs from SQL;
2. If someone wants just `\` at the end but forgot about (1), we will magically work.

Nevertheless, I'm ok with this change.

File: src/Functions/likePatternToRegexp.h [hunk 1]
```diff
@@ -36,13 +43,15 @@ inline String likePatternToRegexp(std::string_view pattern)
                 res += ".";
                 break;
             case '\\':
+                if (pos + 1 == end)
+                    throw Exception(ErrorCodes::CANNOT_PARSE_ESCAPE_SEQUENCE, "Invalid escape sequence");
```
[the remark this answers is not in the document]
Comment (rschu1ze) [reply src/Functions/likePatternToRegexp.h:47]: Done

**Comment by rschu1ze** [comment]
>     2. If someone wants just `\` at the end but forgot about (1), we will magically work.

Let's see what happens now:

```
select 'a\\' LIKE 'a'; // works + returns 0
select 'a\\' LIKE 'a\'; // exception 'Single quoted string is not closed' thrown when un-quoting the pattern
select 'a\\' LIKE 'a\\'; // exception 'Invalid escape sequence at the end of LIKE pattern' during LIKE evaluation
select 'a\\' LIKE 'a\\\'; // exception 'Single quoted string is not closed' thrown when un-quoting the pattern
select 'a\\' LIKE 'a\\\\'; // works + returns 1
```

Looks slightly more correct than before, in particular we don't magically enable patterns which did not work before.

**Review by IhostVlad — APPROVED** [review]

**Comment by rschu1ze** [comment]
There is a ton of exceptions in the fuzzer logs about LIKE predicates (including a few instances of the new exception), but all of them look expected. Should be good for merge once the remaining tests ran.

**Comment by rschu1ze** [comment]
The fuzzer seems to start from test queries newly introduced with the PR, so this would explain the many fuzzer failures.

Other failures are in the "backwards compat stress tests" which run the tests on a CH from early May. This obviously produces many errors because non-const patterns were not available back then.

Merging now (with a slightly uneasy feeling though). In case of problems, we can revert.

**Pull request merged**